### PR TITLE
PMM-7438: fix reading long log lines

### PR DESCRIPTION
**Pull request by michal-kralik**
PMM-7438

Link to the Feature Build: SUBMODULES-3200

### CODE REVIEW

**Comment by BupycHuk** [comment]
@michal-kralik PTAL tests, one of them failed.

**Review by rishat-ishbulatov — APPROVED** [review]

File: managed/services/supervisord/logs.go [hunk 1]
```diff
@@ -302,13 +302,18 @@ func readLog(name string, maxLines int, maxBytes int64) ([]byte, time.Time, erro
...
 	r := ring.New(maxLines)
-	s := bufio.NewScanner(f)
-	for s.Scan() {
-		r.Value = []byte(s.Text() + "\n")
+	reader := bufio.NewReader(f)
+	for {
+		b, err := reader.ReadBytes('\n')
+		if err == io.EOF {
+			break
+		}
```
Comment (PavelKhripkov) [inline managed/services/supervisord/logs.go:310]: Looks like we're losing the last part if source doen't end with new line.

**Review by PavelKhripkov — CHANGES_REQUESTED** [review]

File: managed/services/supervisord/logs.go (hunk 1 above)
Comment (michal-kralik) [reply managed/services/supervisord/logs.go]: Good catch

**Comment by michal-kralik** [comment]
> 1. We're using Scanner in other parts of the code. Are we going to change them as well?
Not as part of this PR.

> 2. Don't we need to add tests to check that bug is really gone?
I personally don't think such test adds value. It only checks if someone uses an incorrect way of reading a file.

**Review by BupycHuk — APPROVED** [review]

File: managed/services/supervisord/logs.go [hunk 2]
```diff
@@ -302,13 +302,24 @@ func readLog(name string, maxLines int, maxBytes int64) ([]byte, time.Time, erro
...
-	for s.Scan() {
-		r.Value = []byte(s.Text() + "\n")
+	reader := bufio.NewReader(f)
+	for {
+		b, err := reader.ReadBytes('\n')
+		if err == io.EOF {
+			// A special case when the last line does not end with a new line
+			if len(b) != 0 {
+				r.Value = b
+				r = r.Next()
```
Comment (PavelKhripkov) [inline managed/services/supervisord/logs.go:312]: Not sure we need `r.Next()` here since we're exiting the loop.

**Comment by PavelKhripkov** [comment]
> > 1. We're using Scanner in other parts of the code. Are we going to change them as well?
> >    Not as part of this PR.
We can create a tech debt ticket then to not forget.


> > 2. Don't we need to add tests to check that bug is really gone?
> >    I personally don't think such test adds value. It only checks if someone uses an incorrect way of reading a file.
Not sure I understood the phrase "an incorrect way of reading a file". Did you check that the described problem has gone after the changes you made?

File: managed/services/supervisord/logs.go (hunk 2 above)
Comment (michal-kralik) [reply managed/services/supervisord/logs.go:312]: We do, otherwise we don't get correct results.

**Comment by michal-kralik** [comment]
> Not sure I understood the phrase "an incorrect way of reading a file". Did you check that the described problem has gone after the changes you made?

Yes, this fixes the issue.
An "incorrect way" would be to use `bufio.NewScanner()` which states it stops on tokens/lines which are too long to fit into a buffer.

**Pull request merged**